server-mux.go

```
@@ -0,0 +1,151 @@
+	peekLen int
+}
+
+func NewMuxConn(c net.Conn) *MuxConn {
+	maxLen := 0
+	for _, m := range defaultHTTPMethods {
+		if maxLen < len(m) {
+			maxLen = len(m) + 1
+		}
+	}
+	return &MuxConn{Conn: c, peekLen: maxLen, buffer: make([]byte, maxLen)}
+}
+
+func (c *MuxConn) peekProtocol() (string, error) {
+	n, err := c.Conn.Read(c.buffer)
+	if n == 0 || err != nil {
+		return "", err
+	}
+	c.reset = true
+	for _, m := range defaultHTTPMethods {
+		if strings.HasPrefix(string(c.buffer), m+" ") {
+			return "http", nil
+		}
+	}
+	return "https", nil
+}
+
+func (c *MuxConn) Read(b []byte) (int, error) {
+	if c.reset {
+		copy(b, c.buffer[c.pointer:])
+		c.pointer += len(b)
+		if c.pointer >= len(c.buffer) {
+			c.reset = false
+			c.pointer = len(c.buffer)
+		}
+		return c.pointer, nil
+	}
+	n, err := c.Conn.Read(b)
+	return n, err
+}
+
+type MuxListener struct {
+	net.Listener
+	config tls.Config
+}
+
+func NewMuxListener(listener net.Listener, certPath, keyPath string) (MuxListener, error) {
+	var err error
+	config := &tls.Config{}
+	if config.NextProtos == nil {
+		config.NextProtos = []string{"http/1.1"}
+	}
+	config.Certificates = make([]tls.Certificate, 1)
+	config.Certificates[0], err = tls.LoadX509KeyPair(certPath, keyPath)
+	if err != nil {
+		return MuxListener{}, err
+	}
+	return MuxListener{Listener: listener, config: *config}, err
+}
+
+func (l MuxListener) Accept() (net.Conn, error) {
+	c, err := l.Listener.Accept()
+	if err != nil {
+		return c, err
+	}
+	cmux := NewMuxConn(c)
+	protocol, err := cmux.peekProtocol()
+	if err == nil && protocol == "https" {
+		if err == nil {
+			return tls.Server(cmux, &l.config), nil
+		}
+	}
+	return cmux, nil
+}
+
+type MuxServer struct {
+	http.Server
+}
+
+func (m *MuxServer) ListenAndServeTLS(certFile, keyFile string) error {
+	listener, err := net.Listen("tcp", m.Server.Addr)
+	mux, err := NewMuxListener(listener, mustGetCertFile(), mustGetKeyFile())
+	if err != nil {
+		fatalIf(err, "Failed to start minio server.")
```

We should return a error here.. since this function is required to.